dirty_cat/gamma_poisson_factorization.py

```diff
@@ -0,0 +1,600 @@
+    #         rescale_W=self.rescale_W,
+    #         gamma_shape_prior=self.gamma_shape_prior,
+    #         gamma_scale_prior=self.gamma_scale_prior)
+    #     self._update_H_dict(unq_X, unq_H)
+    #     _multiplicative_update_w(
+    #         unq_V[lookup], self.W_, self.A_, self.B_,
+    #         unq_H[lookup], self.rescale_W, self.rho_)
+    #     return self
+
+    def _add_unseen_keys_to_H_dict(self, X):
+        """
+        Add activations of unseen string categories from X to H_dict.
+        """
+        unseen_X = np.setdiff1d(X, np.array([*self.H_dict]))
+        if unseen_X.size > 0:
+            unseen_V = self.ngrams_count.transform(unseen_X)
+            if self.add_words:
+                unseen_V2 = self.word_count.transform(unseen_X)
+                unseen_V = sparse.hstack((unseen_V, unseen_V2), format='csr')
+
+            unseen_H = _rescale_h(
+                unseen_V, np.ones((unseen_V.shape[0], self.n_topics)))
+            self._update_H_dict(unseen_X, unseen_H)
+
+    def transform(self, X):
+        """
+        Return the encoded vectors (activations) H of input strings in X.
+        Given the learnt topics W, the activations H are tuned to fit V = HW.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, ) or (n_samples, 1)
+            The string data to encode.
+
+        Returns
+        -------
+        H : 2-d array, shape (n_samples, n_topics)
+            Transformed input.
+        """
+        # Check input data shape
+        X = np.asarray(X)
+        assert X.ndim == 1 or (X.ndim == 2 and X.shape[1] == 1), f"ERROR:\
+        shape {X.shape} of input array is not supported."
+        if X.ndim == 2:
+            X = X[:, 0]
+        # Check if first item has str or np.str_ type
+        assert isinstance(X[0], str), "ERROR: Input data is not string."
+        unq_X = np.unique(X)
+        # Build the n-grams counts matrix V for the string data to encode
+        unq_V = self.ngrams_count.transform(unq_X)
+        if self.add_words: # Add words counts
+            unq_V2 = self.word_count.transform(unq_X)
+            unq_V = sparse.hstack((unq_V, unq_V2), format='csr')
+        # Add unseen strings in X to H_dict
+        self._add_unseen_keys_to_H_dict(unq_X)
+        unq_H = self._get_H(unq_X)
+        # Loop over batches
+        for slice in gen_batches(n=unq_H.shape[0],
+                                 batch_size=self.batch_size):
+            # Given the learnt topics W, optimize H to fit V = HW
+            unq_H[slice] = _multiplicative_update_h(
+                unq_V[slice], self.W_, unq_H[slice],
+                epsilon=1e-3, max_iter=100,
+                rescale_W=self.rescale_W,
+                gamma_shape_prior=self.gamma_shape_prior,
+                gamma_scale_prior=self.gamma_scale_prior)
+        # Store and return the encoded vectors of X
+        self._update_H_dict(unq_X, unq_H)
+        return self._get_H(X)
+
+
+def _rescale_W(W, A, B):
```

Comment:
Is there a reason to pass B? Right now it does not seem to be used.

Reply:
You're right, removed!